### DISCUSSION TITLE
How does "change print resolution" tool work?

### DISCUSSION THREAD

**Discussion by bryanseger** [sn4k3/UVtools#864] [General]
For reasons that deserve their own post, I am using Lychee to slice for a Mini 8KS printer and then using UVtools to resize the slices to fit for a Mini 8K (for some reason, the 8KS has an X resolution of 7536, where the 8K is only 7500). I am ultimately printing on a Mini 8K, and this solution is working fine.

What I'm wondering is, when I use the "change print resolution" tool to remove those 36 extra pixels, how is it doing that? Is it scaling it uniformly across the X, is it cutting off 18 pixels on the far left and 18 pixels on the far right, or is some other way being used?

**Comment by sn4k3** [comment]
8K and 8KS have same pixel size which mean no resize, it will just crop or expand image as needed (If you selected the profile from machine presets correctly).

It will only resize in this conditions: `if (_fixRatio && (newFixedRatioX != 1.0 || newFixedRatioY != 1.0))`

But why not select the right printer on slicer?

**For reference:**
```
Phrozen Sonic Mini 8K | Resolution=7500x3240(px) Display=165x71.28(mm) PixelSize=22x22(µm) Z=180mm
Phrozen Sonic Mini 8K S | Resolution=7536x3240(px) Display=165.792x71.28(mm) PixelSize=22x22(µm) Z=170mm
```

**Reply by bryanseger** [reply]
So in my case, is it simply cropping off 36 pixels from the end, or is it split half & half from left and right, or perhaps it's not doing anything since my print file is nowhere close to the full 7500px width?



It's a complicated reason that I don't yet have the full answers for. The original Mini 8K units shipped with what I'll call the 'old' motherboard. Recent units are shipping with a 'new' motherboard. I went to tweak one of our M8K machines with updated firmware to reduce its allowed overall Z height (for some custom build heads I'm working on). I went to Phrozen's site to download the latest firmware and saw that they're differentiating between 'old' and 'new' motherboards (they even have a different firmware file to download depending on which motherboard you have).

I download the latest firmware for an 'old' machine but noticed that the machine no longer respected wait-before-cure times. I didn't have time to mess with it, so I reverted the machine back to older firmware. However, as I noticed this, I looked at the newest M8K that we have (which has the 'new' motherboard) and noted that it too has the same issue of ignoring the bottom wait-before-cure times and only follows the 'normal' wait-before-cure times. 

In Lychee, for the M8K, there is no bottom wait-before-cure time option, so I typically create that in UVtools (and that works well for all 'old' M8K units). This option now no longer works for the 'new' M8K (unless you think this is a thing you can correct in UVtools). As I was downloading the latest firmware, I noted that it now supports the .PRZ format; however, Lychee doesn't output .PRZ for the M8K, but it does for the M8KS. Lychee also has the bottom wait-before-cure option for the M8KS. So I created a .PRZ test file for a M8KS and plugged that into the 'new' M8K and was surprised to find a resolution error. I had assumed that the M8K and M8KS had the exact same screens, but apparently not. This led me to use your "change print resolution" tool, which worked very well.

Once I had that going, I confirmed that the 'new' M8K does respect both the bottom and normal wait-before-cure times for the .PRZ format, except it does not wait for the very first layer. To get around this, I clone the 0 layer in UVtools and remove all pixels on the new 0 layer, which makes the printer use a legit wait time for the first layer that actually prints. 



Because the latest firmware for the 'old' M8K prevented the printer from respecting the bottom wait-before-cure times, I assume this is on Phrozen's end to fix. I have reached out to them to see if they can address this, or if nothing else, at least provide older versions of the firmware available for download so that users can downgrade if needed. However, if you think there's something you can change in UVtools to get around this, that would be amazing. I also wonder if part of this is on Lychee since they do have a bottom wait-before-cure setting for the M8KS, but they do not for the M8K. 

It's a long-winded answer but I uncovered a lot today. I believe I have a work-around but it would be nice if I didn't have to do it this way! Thanks again Tiago!

**Reply by sn4k3** [reply]
> So in my case, is it simply cropping off 36 pixels from the end, or is it split half & half from left and right, or perhaps it's not doing anything since my print file is nowhere close to the full 7500px width?

In your case it create a empty image with new resolution, copy and paste the model area (Bounding rectangle) into new image absolute center, pixels are preserved as they are. That tool also update resolution and display information to file.

What a mess with the firmwares... Can't you edit resolution in Lychee?
Note that CTB and PZR are different formats.

> To get around this, I clone the 0 layer in UVtools and remove all pixels on the new 0 layer, which makes the printer use a legit wait time for the first layer that actually prints.

If you apply UVtools suggestions for wait time, it will take care of the empty layer for you.

> I have reached out to them to see if they can address this

Good luck with that.

**Reply by bryanseger** [reply]
To respond to your notes/question:

- No, resolution cannot be changed in Lychee (and I've got Pro!). I can change print volume, but not resolution. 
- Yes, I switched over to PRZ at least to try it out (because I noted on Phrozen's site that the latest firmware for M8K supports PRZ).
- The UVtools suggestion does not apply the empty layer for a PRZ file (only CTB, I think). Maybe a PRZ on a M8KS or the Mega 8KS doesn't need the empty layer, but it does on the M8K.
- Yea...Phrozen's support believes I did not format my USB correctly, and they're not understanding the issue...so I don't expect their help to go any further unless they can provide older firmware files.


I did some more testing this morning:

- A PRZ straight from Lychee respects both the bottom and normal wait-before-cure times, except it does not wait for the first layer (as soon as the build head zeros out, it begins the print).
- I took that same file and manually created the empty layer in UVtools. This works on the printer as I want it to, except the first true layer is one layer higher than the empty layer, despite UVtools saying they're both at the same Z height. It's like the M8K (with new mainboard/firmware) is only looking at global variables instead of layer variables. I'm not sure if you can do anything about that or not.
- I then took the same PRZ file from Lychee and applied the "wait time before cure" suggestion in UVtools. This expectedly did not create the empty layer, so I manually created it. This file has the same Z issue as above, where the first true layer is one layer higher than the empty layer, despite UVtools showing them both at the same Z height. But also, the transition of wait-before-cure times (that the suggestion applied) were not respected by the printer. Instead of stepping down from 20s to 5s over ~10 layers, it jumped straight from 20s to 5s at the same transition point from bottom layers to normal layers (acting essentially as what I have set in Lychee).

Based on this testing, it seems that the new M8K and new firmware for it makes it no longer respect the layer variables and only the global variables. Maybe something will jump out to you that you think you could fix, but I think this is on Phrozen's side.

**Reply by sn4k3** [reply]
> No, resolution cannot be changed in Lychee (and I've got Pro!). I can change print volume, but not resolution.

Also not possible to create a custom printer from scratch?
To facilitate your workflow you can prepare a .uvtop file to drag and drop into uvtools after open the file and it will apply the resolution and display for you.


> The UVtools suggestion does not apply the empty layer for a PRZ file (only CTB, I think). Maybe a PRZ on a M8KS or the Mega 8KS doesn't need the empty layer, but it does on the M8K.

You are right, I only had that option for CTB, I will add this format too.
They ignore all waits from first layer because maybe it never make sense for them to wait in first layer as nothing is printed yet, but they won't change this so soon. The problem of need a first empty layer is that it require the bad supported per layer settings flag on... You can read more here #859 the problems that's causing (other printer and format) still this flag is being an annoyance among formats.

> I took that same file and manually created the empty layer in UVtools. This works on the printer as I want it to, except the first true layer is one layer higher than the empty layer, despite UVtools saying they're both at the same Z height. It's like the M8K (with new mainboard/firmware) is only looking at global variables instead of layer variables. I'm not sure if you can do anything about that or not.

After save  file, in info properties look for: "PerLayerSettings", it must show as **True** or **1** for that to work as requesite.
But as you described I think Its a firmware issue.
Here what I think: PZR share the GOO format, but it have a difference, layers do not support X grey compression. But if not the only difference and PRZ also does not take into account the "PerLayerSettings" flag and just print globally settings? I think in past I saw a issue stating something similar to you, so right now this is my guess, while PZR is GOO format it does not support some features. (Capped format)
